File: engine/wal_test.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package engine

import (
	"os"
	"path/filepath"
	"testing"
)

func TestWal_MigratesLegacyDataLog(t *testing.T) {
	dir := t.TempDir()
	legacyPath := filepath.Join(dir, logFileName)
	if err := os.WriteFile(legacyPath, []byte{0, 0, 0, 0}, 0o644); err != nil {
		t.Fatal(err)
	}

	log, err := OpenDataLog(dir, nil, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer log.Close()

	if _, err := os.Stat(legacyPath); !os.IsNotExist(err) {
		t.Fatal("expected legacy data.log migrated away")
	}
	manifestPath := filepath.Join(dir, walDirName, walManifestName)
	if _, err := os.Stat(manifestPath); err != nil {
		t.Fatalf("expected manifest at %s: %v", manifestPath, err)
	}
}

func TestWal_GlobalLSNAcrossSegmentRotation(t *testing.T) {
	dir := t.TempDir()
	const segSize = 256
	log, err := OpenDataLog(dir, nil, segSize)
	if err != nil {
		t.Fatal(err)
	}
	defer log.Close()

	payload := encodeRecord(Record{Type: RecordSet, XID: 1, Key: []byte("k"), Value: []byte("v")})
	frameLen := int(frameSize(len(payload)))

	var firstOff int64 = -1
	for i := 0; i < 20; i++ {
		off, err := log.AppendEncoded(payload, false)
		if err != nil {
			t.Fatal(err)
		}
		if firstOff < 0 {
			firstOff = off
		}
	}
	if log.SegmentCount() < 2 {
		t.Fatalf("expected segment rotation with small segment size, got %d segments", log.SegmentCount())
	}

	val, err := log.ReadValueAt(firstOff, 1)
	if err != nil || string(val) != "v" {
		t.Fatalf("read from first segment after rotation: err=%v val=%q", err, val)
	}
	if lastOff := log.WriteOffset(); lastOff <= firstOff+int64(frameLen) {
		t.Fatalf("expected global LSN to grow across segments, first=%d last=%d", firstOff, lastOff)
	}
}

func TestWal_ReadLogRangeCrossesSegmentBoundary(t *testing.T) {
	dir := t.TempDir()
	const segSize = 128
	log, err := OpenDataLog(dir, nil, segSize)
	if err != nil {
		t.Fatal(err)
	}

	payload := encodeRecord(Record{Type: RecordSet, XID: 1, Key: []byte("k"), Value: []byte("value")})
	for i := 0; i < 15; i++ {
		if _, err := log.AppendEncoded(payload, false); err != nil {
			t.Fatal(err)
		}
	}
	head := log.WriteOffset()
	data, end, err := log.ReadLogRange(0, head)
	if err != nil {
		t.Fatal(err)
	}
	if len(data) == 0 || end <= 0 {
		t.Fatalf("expected cross-segment read, len=%d end=%d", len(data), end)
	}
	frames, err := validateFrames(data)
	if err != nil {
		t.Fatalf("validate cross-segment batch: %v", err)
	}
	if len(frames) == 0 {
		t.Fatal("expected frames from cross-segment read")
	}
	log.Close()

	db, err := Open(dir, Options{WalSegmentSize: segSize})
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	if db.log.SegmentCount() < 2 {
		t.Fatalf("expected persisted segments after reopen, got %d", db.log.SegmentCount())
	}
}

func TestWal_ReplayPreservesGlobalOffsets(t *testing.T) {
	dir := t.TempDir()
	opts := Options{WalSegmentSize: 200}

	db, err := Open(dir, opts)
	if err != nil {
		t.Fatal(err)
	}
	tx := db.NewTransaction(true)
	if err := tx.Put([]byte("k"), []byte("v")); err != nil {
		t.Fatal(err)
	}
	if err := tx.Commit(); err != nil {
		t.Fatal(err)
	}
	var storedOffset int64
	db.index.ForEachKey(func(_ []byte, chain []indexVersion) {
		if len(chain) > 0 {
			storedOffset = chain[0].offset
		}
	})
	db.Close()

	db2, err := Open(dir, opts)
	if err != nil {
		t.Fatal(err)
	}
	defer db2.Close()

	val, err := db2.log.ReadValueAt(storedOffset, 1)
	if err != nil || string(val) != "v" {
		t.Fatalf("replay offset read failed: err=%v val=%q offset=%d", err, val, storedOffset)
	}
}

func TestWal_IsFrameBoundaryAtSegmentBase(t *testing.T) {
	dir := t.TempDir()
	log, err := OpenDataLog(dir, nil, 128)
	if err != nil {
		t.Fatal(err)
	}
	defer log.Close()

	payload := encodeRecord(Record{Type: RecordSet, XID: 1, Key: []byte("a"), Value: []byte("1")})
	for i := 0; i < 10; i++ {
		if _, err := log.AppendEncoded(payload, false); err != nil {
			t.Fatal(err)
		}
	}
	if log.SegmentCount() < 2 {
		t.Fatal("expected rotation")
	}
	// Second segment base should be a valid frame boundary.
	base := log.segments[1].baseLSN
	if !log.IsFrameBoundary(base) {
		t.Fatalf("expected frame boundary at segment base %d", base)
	}
}

File: rust/crates/turnstone-engine/tests/wal_test.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use turnstone_engine::hashindex::IndexExt;
use turnstone_engine::{
    encode_record, frame_size, replay_into_mem, validate_frames, DataLog, Record, RecordType,
};
use turnstone_engine::wal::wal_segment_file_name;

#[test]
fn wal_creates_fresh_manifest_on_open() {
    let dir = tempfile::tempdir().unwrap();
    let log = DataLog::open(dir.path(), 0).unwrap();
    let manifest_path = dir.path().join("wal").join("manifest.json");
    assert!(manifest_path.exists());
    let seg_path = dir.path().join("wal").join(wal_segment_file_name(1));
    assert!(seg_path.exists());
    log.close().unwrap();
}

#[test]
fn wal_global_lsn_across_segment_rotation() {
    let dir = tempfile::tempdir().unwrap();
    const SEG_SIZE: i64 = 256;
    let log = DataLog::open(dir.path(), SEG_SIZE).unwrap();
    let payload = encode_record(&Record {
        ty: RecordType::Set,
        xid: 1,
        key: b"k".to_vec(),
        value: b"v".to_vec(),
    });
    let frame_len = frame_size(payload.len()) as i64;
    let mut first_off = -1i64;
    for _ in 0..20 {
        let off = log.append_encoded(&payload, false).unwrap();
        if first_off < 0 {
            first_off = off;
        }
    }
    assert!(
        log.segment_count() >= 2,
        "expected rotation, got {} segments",
        log.segment_count()
    );
    let val = log.read_value_at(first_off, 1).unwrap();
    assert_eq!(val, b"v");
    let last = log.write_offset();
    assert!(last > first_off + frame_len);
    log.close().unwrap();
}

#[test]
fn wal_read_log_range_crosses_segment_boundary() {
    let dir = tempfile::tempdir().unwrap();
    const SEG_SIZE: i64 = 128;
    let log = DataLog::open(dir.path(), SEG_SIZE).unwrap();
    let payload = encode_record(&Record {
        ty: RecordType::Set,
        xid: 1,
        key: b"k".to_vec(),
        value: b"value".to_vec(),
    });
    for _ in 0..15 {
        log.append_encoded(&payload, false).unwrap();
    }
    let head = log.write_offset();
    let (data, end) = log.read_log_range(0, head).unwrap();
    assert!(!data.is_empty() && end > 0);
    let frames = validate_frames(&data).unwrap();
    assert!(!frames.is_empty());
    log.close().unwrap();

    let log2 = DataLog::open(dir.path(), SEG_SIZE).unwrap();
    assert!(log2.segment_count() >= 2);
    log2.close().unwrap();
}

#[test]
fn wal_replay_preserves_global_offsets() {
    let dir = tempfile::tempdir().unwrap();
    let log = DataLog::open(dir.path(), 200).unwrap();
    let begin = encode_record(&Record {
        ty: RecordType::Begin,
        xid: 1,
        key: vec![],
        value: vec![],
    });
    let set = encode_record(&Record {
        ty: RecordType::Set,
        xid: 1,
        key: b"k".to_vec(),
        value: b"v".to_vec(),
    });
    let commit = encode_record(&Record {
        ty: RecordType::Commit,
        xid: 1,
        key: vec![],
        value: vec![],
    });
    log.append_encoded(&begin, false).unwrap();
    let set_off = log.append_encoded(&set, false).unwrap();
    log.append_encoded(&commit, true).unwrap();
    log.close().unwrap();

    let log2 = DataLog::open(dir.path(), 200).unwrap();
    let (index, _, _) = replay_into_mem(&log2, false).unwrap();
    let chain = index.get_chain(b"k");
    let stored = chain[0].offset;
    assert_eq!(stored, set_off);
    let val = log2.read_value_at(stored, 1).unwrap();
    assert_eq!(val, b"v");
    log2.close().unwrap();
}

#[test]
fn wal_is_frame_boundary_at_segment_base() {
    let dir = tempfile::tempdir().unwrap();
    let log = DataLog::open(dir.path(), 128).unwrap();
    let payload = encode_record(&Record {
        ty: RecordType::Set,
        xid: 1,
        key: b"a".to_vec(),
        value: b"1".to_vec(),
    });
    for _ in 0..10 {
        log.append_encoded(&payload, false).unwrap();
    }
    assert!(log.segment_count() >= 2);
    let segs = log.segments();
    let base = segs[1].1;
    assert!(log.is_frame_boundary(base), "expected boundary at {base}");
    log.close().unwrap();
}

#[test]
fn recovery_uncommitted_not_in_index() {
    let dir = tempfile::tempdir().unwrap();
    let log = DataLog::open(dir.path(), 4096).unwrap();
    let begin = encode_record(&Record {
        ty: RecordType::Begin,
        xid: 1,
        key: vec![],
        value: vec![],
    });
    let set = encode_record(&Record {
        ty: RecordType::Set,
        xid: 1,
        key: b"crash_key".to_vec(),
        value: b"before_commit".to_vec(),
    });
    log.append_encoded(&begin, false).unwrap();
    log.append_encoded(&set, false).unwrap();
    log.close_active_file_for_test();
    log.close().unwrap();

    let log2 = DataLog::open(dir.path(), 4096).unwrap();
    let (index, clog, _) = replay_into_mem(&log2, true).unwrap();
    assert!(index.get_chain(b"crash_key").is_empty());
    assert_eq!(clog.clog_len(), 0);
    log2.close().unwrap();
}
